Add coverage to project reports.

kinow opened the pull request:
Hi,

This is based on a [pull request submitted for Cylc](https://github.com/cylc/cylc/pull/2766), but with a few enhancements (which will be applied there too now).

The current [coverage reported in Codacy is **3%**](https://app.codacy.com/project/kinow/rose/dashboard), but it could be due to some bug in this pull request. Or maybe not all the code can be covered, and so it would be just a matter of ignoring more files... or maybe it is due to some configuration from Codacy that I forgot to change.

I finished this change yesterday, and have been playing with Codacy and Travis until now, trying to tweak the settings to see if there was anything obvious that I was missing. But perhaps somebody else would be interested in having a look at it? (if you compare my/rose Codacy dashboard, there are a small improvement in the metrics due to myself adding more files to be ignored... right now only lib/python/{rose|rosie} are being analyzed - I think).

Before merging it, someone with admin access to the project's Codacy account would have to [follow the Codacy docs](https://support.codacy.com/hc/en-us/articles/207279819-Coverage) to enable Travis integration, and also customize what files are ignored in the analysis (e.g. isodatetime, demo, etc, and other folders may be analyzed and considered for coverage statistic).

Cheers
Bruno

matthewrmshin commented:
HI @kinow Looking at the list of files with no test coverage, I think there is definitely something not right somewhere. I have expected many of those files to have at least some coverage even under the *lean* Rose test battery on Travis CI. (We run a fatter test in our site environment to cater for remote host settings etc.)

kinow commented:
Hi @matthewrmshin that's what I thought too. It can't be 3%, unless Codacy is perhaps considering the total lines of code (including documentation, tests, etc), or something else is wrong with my settings. I am still validating the settings for Cylc (with [47%](https://app.codacy.com/project/kinow/cylc/dashboard?branchId=8624609) right now), and then will try to understand better why the final number here was so low.

kinow commented:
Around 20% coverage. I think that's correct, mainly due to some GUI code (which can be omitted/ignored via the `omit` entry in `coveragerc`. Cheers, Bruno

matthewrmshin commented:
Looking at some individual files (that should be well tested), the reported coverage is still way too low.

kinow commented:
If you can point to a few of these files, I can investigate why they are either not appearing, or why their coverage is too low.

matthewrmshin commented:
This one for example should be well tested:
https://codecov.io/gh/metomi/rose/src/master/lib/python/rose/env_cat.py

kinow commented:
Fixed!!!

And we have [**84.92%**](https://codecov.io/gh/kinow/rose/branch/report-test-coverage) covered.

The problem was that `source` worked with directories, but it wasn't happy about `rose` and `rosie` for some reason.

So I found a user who reported a workaround by using `include`. Tried that today, and didn't work.

Well, turns out `source` accepts `./lib/python`, but `include` uses that value as `regex` apparently and that doesn't work.

Replaced by `*lib/python/rose*` and same for rosie, and now it's working fine. Found it after searching the internet, and by looking at the debug output.

Well done for the devs for such a good test coverage. It seems quite easy to reach 95%, just by writing tests for DAO's and other simple classes.

Ready for review again.

Thanks

matthewrmshin commented on .travis/cover.py:
```diff
@@ -0,0 +1,31 @@
+def main():
+    call(
+        'rose test-battery --state=save -j 5 || (echo -e "\n\nRerunning Failed Tests...\n\n"; rose test-battery -v --state=save,failed -j 5)',
+        shell=True)
```
To reduce the shell syntax and to return with the correct exit code, perhaps we can try (not tested):

```python
command = ['rose', 'test-battery', '-j', '5']
if call(command + ['--state=save']):
    # Non-zero return code
    import sys
    sys.stderr.write('\n\nRerunning Failed Tests...\n\n')
    # Exit with final return code
    sys.exit(call(command + ['--state=save,failed', '-v']))
```

kinow replied:
Great idea @matthewrmshin ! Copied your code as-is, first it failed, but then the next run (where I squashed commits) it re-ran the tests and it worked. Before the script execution would always report success, passing the build :disappointed: 

Going to fix the Cylc pull request with this neat approach. Thanks Matt!

oliver-sanders commented:
It would be good to get this branch square with it's Cylc equivalent, are we going to copy across the `.codecov.yml`?

kinow commented:
Well noted @oliver-sanders ! Forgot to add that file. Just edited the last commit to include the file based on Cylc's. But set the minimum to 84%, allowing 1% decrease in pull requests. Will leave it up to you guys to decide the best values.

At work, when I pushed this change, I had to trigger the Travis build a couple of times until it succeeded.

Now I changed only the `.codecov.yml`, but some tests are randomly failing. Tried 6 builds, triggered the last one now. Are the tests normally this unstable, or is it something I am missing in my pull request, or perhaps some change that I missed in .travis.yml?

matthewrmshin commented:
Test will be fixed by #2257. Rose test battery is normally quite stable.

kinow commented:
@matthewrmshin had lost track of this one, thanks for the remainder.

@oliver-sanders 

>Since you've found a way of combining the coverage reports of multiple Travis-CI jobs in Cylc, would it be possible to go back to running the functional and documentation tests in separate jobs. Useful, particularly for documentation only changes.

Would you like to take another look and see what else needs changing? I can update it before the weekend.

oliver-sanders commented:
> Would you like to take another look and see what else needs changing?

@kinow I can make this change in #2222 if necessary but I don't possess your expertise regards combining coverage reports in different jobs. This diff is is what I was thinking (restores parallelism and fits in with #2222), would the coverage still work?

```diff
diff --git a/.travis.yml b/.travis.yml
index 617363cf..3cc9a099 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -34,19 +34,22 @@ install:
     - sudo apt-get update
     - sudo apt-get install -y subversion
 
-script:
-    - export COVERAGE_PROCESS_START="${TRAVIS_BUILD_DIR}/.coveragerc"
-    # ./bin/rose will change PYTHONPATH, so we cannot rely on that for sitecustomize.py
-    # but instead we can move the sitecustomize.py to the place used as PYTHONPATH
-    - cp "${TRAVIS_BUILD_DIR}/.travis/sitecustomize.py" ./lib/python
-    - export PYTHONPATH="${TRAVIS_BUILD_DIR}/.travis"
-    - coverage run .travis/cover.py
-    - rm ./lib/python/sitecustomize.py
-
-after_script:
-    - rose make-docs --venv --dev --strict clean linkcheck doctest html slides latexpdf
-
 after_success:
     - coverage combine --append
     - coverage xml --ignore-errors
     - bash <(curl -s https://codecov.io/bash)
+
+jobs:
+    include:
+    - name: "Test Battery"
+      script:
+        - export COVERAGE_PROCESS_START="${TRAVIS_BUILD_DIR}/.coveragerc"
+        # ./bin/rose will change PYTHONPATH, so we cannot rely on that for sitecustomize.py
+        # but instead we can move the sitecustomize.py to the place used as PYTHONPATH
+        - cp "${TRAVIS_BUILD_DIR}/.travis/sitecustomize.py" ./lib/python
+        - export PYTHONPATH="${TRAVIS_BUILD_DIR}/.travis"
+        - coverage run .travis/cover.py
+        - rm ./lib/python/sitecustomize.py
+    - name: "Documentation"
+      script:
+        - rose make-docs --venv --dev --strict clean linkcheck doctest html slides latexpdf
```

kinow commented:
Hi @oliver-sanders ,

I think your patch looks good, and I thought it would work, but only one way of knowing for sure :slightly_smiling_face: 

https://travis-ci.org/kinow/rose/jobs/461563649, failing due to docs, but successfully uploaded coverage report to Codecov: https://codecov.io/gh/kinow/rose/branch/report-test-coverage

Coverage at 85%. Will submit another pull for the docs issues if it's something simple.

Branch updated, ready for review again :grin: 

Cheers
Bruno

kinow commented:
`git` is beautiful. `git rebase -i HEAD~2`, remove the last commit, then `git rebase upstream/master` and everything works so nicely. And think that just had to spend a few minutes struggling with a local SVN merge :rage: 

And thanks for @matthewrmshin 's fix for the line length, this PR is again ready for review. Travis happy :+1:
